ambry-messageformat/src/main/java/com.github.ambry.messageformat/CompositeBlobInfo.java

```diff
@@ -37,6 +44,63 @@ public CompositeBlobInfo(int chunkSize, long totalSize, List<StoreKey> keys) {
     this.chunkSize = chunkSize;
     this.totalSize = totalSize;
     this.keys = keys;
+    metadataContentVersion = MessageFormatRecord.Metadata_Content_Version_V2;
+    long last = 0;
+    for (StoreKey key : keys) {
+      offsets.add(last);
+      chunkMetadataList.add(new ChunkMetadata(key, last, Math.min(chunkSize, totalSize - last)));
+      last += chunkSize;
+    }
+  }
+
+  /**
+   * Construct a {@link CompositeBlobInfo} object.
+   * @param keysAndContentSizes list of store keys and the size of the data content they reference
+   */
+  public CompositeBlobInfo(List<Pair<StoreKey, Long>> keysAndContentSizes) {
+    this.chunkSize = -1;
+    this.keys = new ArrayList<>();
+    long last = 0;
+    metadataContentVersion = MessageFormatRecord.Metadata_Content_Version_V3;
+    for (Pair<StoreKey, Long> keyAndContentSize : keysAndContentSizes) {
+      StoreKey key = keyAndContentSize.getFirst();
+      keys.add(key);
+      offsets.add(last);
+      chunkMetadataList.add(new ChunkMetadata(key, last, keyAndContentSize.getSecond()));
+      last += keyAndContentSize.getSecond();
+    }
+    this.totalSize = last;
+  }
+
+  /**
+   * Returns the keys (along with their data content size and offset relative to the total
+   * composite blob) of the chunks related to the given byte range of the entire composite
+   * blob
+   * @param start inclusive starting byte index
+   * @param end inclusive ending byte index
+   * @return the list of {@link ChunkMetadata} that lie upon the inclusive range
+   * between 'start' and 'end'
+   */
+  public List<ChunkMetadata> getStoreKeysInByteRange(long start, long end) {
```

Comment:
agreed, actually writing that now.  `CompositeBlobInfo` didn't have a unit test before because it was a POJO class but with this method it should be tested.

Reply:
Can you explain "0 byte chunks in the middle of the blob"?

Reply:
Sure. So I was thinking that since we allow 0 byte simple blob uploads, they are valid candidates for stitching. So, someone could potentially create a blob with chunk sizes like: 25, 30, 0, 0, 70, 0, 5

If we call this method with a `start` offset of 30, would the range start from chunk 1 or 2?

Reply:
As discussed in person, going to change the code so that stitched blobs will not allow blobs of 0 length to be stitched (will throw an exception if that happens)